### PR TITLE
Fix #7926: Redo parameter aliasing optimization

### LINKED ISSUE #7926: Illegal access error when constructor parameter has same name and more specific type
## minimized code

```Scala
trait Foo {
  def parent: Option[Foo]
  def getParent: Option[Foo] = parent
}

class Bar(val parent: Option[Foo]) extends Foo
class Qux(parent: Option[Qux]) extends Bar(parent)

object Qux {
  val qux: Qux = new Qux(None)

  def main(args: Array[String]): Unit = println(qux.getParent)
}
```

<details>
<summary>This code compiles, but crashes at runtime.</summary>

```scala
Exception in thread "main" java.lang.IllegalAccessError: Qux.parent()Lscala/Option;
        at Foo.getParent(Foo.scala:3)
        at Bar.getParent(Foo.scala:6)
        at Qux$.main(Foo.scala:12)
        at Qux.main(Foo.scala)
```
</details>

## expectation

It should print `None`, like Scala 2 does. Note that changing `Qux`'s constructor parameter name to something other than `parent` fixes the issue.

### PR DESCRIPTION
**Pull request by odersky**
- Move main optimization after unpickling, so that changes
   to parameters in superclasses do not break Tasty compatibility.
 - Replace internal Alias annotation by SuperParamAlias flag
 - Don't generate illegal private overrides
 - Make the transformation more robust under separate compilation
   and wrt compilation order

As a separate issue, make sure the user cannot write the sort of illegal "override" of a public with a private method that the compiler generated for this example before.

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/transform/PostTyper.scala [hunk 1]
```diff
@@ -98,6 +98,30 @@ class PostTyper extends MacroTransform with IdentityDenotTransformer { thisPhase
 
     def isCheckable(t: New): Boolean = !inJavaAnnot && !noCheckNews.contains(t)
 
+    /** Mark parameter accessors that are aliases of like-named parameters
+     *  in their superclass with SuperParamAlias.
+     *  This info is used in phase ParamForwarding
+     */
+    private def forwardParamAccessors(impl: Template)(using Context): Unit = impl.parents match
+      case superCall @ Apply(fn, superArgs) :: _ if superArgs.nonEmpty =>
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/transform/PostTyper.scala]: This should exclude super-calls to traits

File: compiler/src/dotty/tools/dotc/transform/PostTyper.scala [hunk 2]
```diff
@@ -98,6 +98,30 @@ class PostTyper extends MacroTransform with IdentityDenotTransformer { thisPhase
...
+        fn.tpe.widen match
+          case MethodType(superParamNames) =>
+            for case stat: ValDef <- impl.body do
```
[the remark this answers is not in the document]
Comment (odersky) [reply compiler/src/dotty/tools/dotc/transform/PostTyper.scala]: No, it's only the supercall of the first template parent

File: compiler/src/dotty/tools/dotc/transform/PostTyper.scala (hunk 1 above)
Comment (odersky) [reply compiler/src/dotty/tools/dotc/transform/PostTyper.scala]: Can't happen. First super call is always to a class.

File: compiler/src/dotty/tools/dotc/transform/PostTyper.scala [hunk 3]
```diff
@@ -98,6 +98,30 @@ class PostTyper extends MacroTransform with IdentityDenotTransformer { thisPhase
...
+                 && !sym.info.isInstanceOf[ExprType] // val-parameters cannot be call-by name, so no need to try to forward to them
+              then
+                val idx = superArgs.indexWhere(_.symbol == sym)
+                if idx >= 0 && superParamNames(idx) == stat.name then
+                  // Supercall to like-named parameter.
+                  // Having it have the same name is needed to maintain correctness in presence of subclassing
+                  // if you would use parent param-name `a` to implement param-field `b`
+                  // overriding field `b` will actually override field `a`, that is wrong!
+                  typr.println(i"super alias: ${sym.showLocated}")
+                  sym.setFlag(SuperParamAlias)
```
[the remark this answers is not in the document]
Comment (odersky) [reply compiler/src/dotty/tools/dotc/transform/PostTyper.scala]: Good point. I changed that.

**Pull request merged**